File: javascript/lesson 1/binarySearchLeft.js
function binarySearchLeft(nums, target) {
    let lo = 0;
    let hi = nums.length - 1;

    while (lo <= hi){
        let mid = lo + Math.floor((hi - lo + 1) / 2);        
        if (nums[mid] > target) {                
            hi = mid - 1;
        } else if (nums[mid] < target) {
            lo = mid + 1;
        } else {                
            if (mid === 0 || nums[mid - 1] < target) return mid;
            hi = mid - 1;
        }
    }

    return -1;
}

File: ruby/lesson 1/binary_search_left.rb
def binary_search_left(nums, target)
    lo = 0
    hi = nums.length - 1

    while lo <= hi
        mid = lo + (hi - lo) / 2
        next hi = mid - 1 if nums[mid] > target
        next lo = mid + 1 if nums[mid] < target
        return mid if mid.zero? || nums[mid - 1] < target
        hi = mid - 1
    end

    -1
end
